fixes 2764: settings defined in seetings.yaml file are now read-only

dmitri-d opened the pull request:
I'm not entirely sure what to do about the current behaviour for settings defined in the configuration file -- they are used to set default values. 
- Should this behaviour be preserved? I think it's awkward.
- If we want to use the configuration file to set default values, should those be updated when configuration file changes? I think so, as this would be less confusing than current semantic (default value can only be set once).

dmitri-d committed:
fixes 2764: settings defined in seetings.yaml file are now read-only

domcleal commented:
Yes, just updating the value (rather than the default) is better I think, particularly combined with the read-only functionality.

The aim is to be able to deploy oauth_\* settings into settings.yaml from the installer and have them populate in the UI and be active.  The trouble at the moment, as you say, is this only seems to happen once - which means it only works if the settings are there before the first db:migrate call.

dmitri-d commented:
Ok, I'll make the change to update the value based on the config file (Do we need the concept of a default value at all now?). 

This change partially solves the problem you described: values defined in the config file now always take precedence (and are read-only).

domcleal commented:
Default value is still important for the default values defined in app/models/settings/*.rb I think.

Ah yes, I hadn't noticed it was updating self.value on find.

dmitri-d commented:
I guess I'm not entirely sure about logic for setting value updates when config file changes. We agreed that the setting value (persisted in the db) needs to change. Do we change the default value too? 

That feels awkward to me: if you want to change the default value you need to update the config file (which also wipes out the current value), restart the app, (optionally) disable the setting in the config file, restart again. Perhaps we need to make the default value editable via web ui?

domcleal commented:
I'd have the default value only coming from the default values hard coded into the STI models, then use SETTINGS to populate the persisted _value_ instead of the default and to make it read only.  The default value stays untouched, it's whatever the STI model defined in the first place.  Does that make sense?

dmitri-d commented:
Yes, it does. I was confused about how SETTINGS affect Setting's default and current values. We have an accord now!

domcleal commented:
Great :)  Yeah, I don't think it makes much sense at the moment - or at least, not how Foreman's currently used.  This brings us something more useful I think.

lzap commented on app/helpers/settings_helper.rb:
```diff
@@ -1,6 +1,12 @@
+    if setting.readonly?
+      return readonly_field(
+          setting, :value,
+          {:title => _("This setting is defined in the configuration file and is read-only."), :helper => :show_value})
```
Maybe to immediately say in which file to direct users there.

Thanks for picking this one up - was a bummer.

lzap commented on app/helpers/application_helper.rb:
```diff
@@ -310,6 +310,20 @@ def gravatar_url(email, default_image)
+      h(value)
+    end
+  end
+
```
Is this really useful in the global context?

dmitri-d replied:
dunno; I suppose I could move it into settings_helper. This location is more obvious though, methinks.

dmitri-d commented:
@domcleal, @lzap: ping? This change updates settings when configuration file changes.

dmitri-d commented:
[test]

lzap commented:
Works for me. @domcleal if you haven't test that yet, I can test. Tests haven't been initiated yet, trying again: [test]

domcleal commented on test/unit/setting_test.rb:
```diff
@@ -113,6 +113,31 @@ def test_set_method_uses_values_from_SETTINGS
   end
 
+  def test_attributes_in_SETTINGS_are_readonly
+    setting_name = "foo_#{::Random.new.rand(100000).to_s}"
```
Just use rand() (Kernel#rand) as Random isn't present on 1.8

domcleal commented on app/models/setting.rb:
```diff
@@ -200,6 +201,14 @@ def clear_cache
+  def settings_file_override
+    if !new_record? && SETTINGS.key?(name.to_sym)
+      self.value = SETTINGS[name.to_sym]
+      update_column(:value, SETTINGS[name.to_sym])
```
Why update_column instead of update_attribute - doesn't update_column write to the DB each time?  Actually, update_attribute also skips validation, which I don't think is a good idea.  Maybe just set it and save?

dmitri-d replied:
Can't. #save uses callbacks, which in turn triggers this very callback. I could only do update if the attribute is dirty though.

domcleal replied:
Ack, I guess that's best then.. I'd prefer to leave validation on if possible, so we catch bad data in settings.yaml early.

domcleal commented:
The only thing about updating the setting value in the find callback is that we duplicate the logic in self.set to pull from SETTINGS.  I now see how this works (sorry for the earlier confusion), as self.set is called from the individual STI models, which then call create!.  Since the setting already exists, we don't update the value if it's supplied (from SETTINGS) - only the default.  On the plus side, at least the SETTINGS and readonly logic are in one place, which is good.  I'm just babbling though, it's probably OK where it is.

@lzap sure, thanks.

dmitri-d commented:
@domcleal: now with validations and dirty attribute check. Also: Did the CI get stuck?

domcleal commented:
CI checks every 15 minutes for new PR changes to test, give it time :)

domcleal commented:
CI's running n ow.

I just tried this out and it blew up when setting `:Parametrized_Classes_in_ENC: false` in settings.yaml and launching `rails s`.  It looks like the update that `Setting.create` does for the default value (even though it shouldn't have changed) is broken by making the whole record readonly.  Could you take a look please?

dmitri-d commented:
@domcleal, @lzap: had to change a chunk of functionality: I removed overridden create and create! methods, as they changed the semantic of the original call, which was part of the problem. Instead I added init_on_startup! (also see load_defaults methods in sub-classes) method that performs required create or update operations as needed without changing any of the ActiveRecord methods.

Unfortunately I had to drop validation on the "value" field when it's being updated from init_on_startup - I don't think there's a clean way to do that with the current approach.

What would work, however, is declaring "name", "default", "description", "category", and "settings_type" fields into read-only ones, and rely on migrations for any subsequent changes for those fields. This would result in a more consistent (semantics-wise) and simpler implementation of Setting.

lzap commented on app/models/setting/auth.rb:
```diff
@@ -7,23 +7,19 @@ def self.load_defaults
+        Setting::Auth.init_on_startup!('ssl_client_verify_env', N_('Environment variable containing the verification status of a client SSL certificate'), 'SSL_CLIENT_VERIFY')
+        Setting::Auth.init_on_startup!('signo_sso', N_('Use Signo SSO for login'), false)
+        Setting::Auth.init_on_startup!('signo_url', N_('Signo SSO url'), "https://#{fqdn}/signo")
     end
```
Hmm looks better now, but we are breaking plugin compatibility here. E.g. discovery does use self.set :-(

ohadlevy replied:
:+1: on what @lzap said, we will break any plugin that uses settings, I would prefer to avoid it if possible.

dmitri-d replied:
Do we have a lot of plugins? If it's a change in one or two, it's not a real issue. I don't think I can make this change fully backwards-compatible, and it's worth fixing current Settings api.

I can introduce a temporary (see below) work around this by keeping Setting.set() and replace "update certain fields on create" semantics by simply ignoring updates on create altogether to give plugin authors time to update their code. Ultimately this will have to go though, as the suggested fix introduces a problem with duplicate settings beings silently discarded.

ohadlevy replied:
we have many, including many that we are not aware of (e.g. developed in house by someone).

dmitri-d replied:
Ok, what are the thoughts re: a temporary workaround?

GregSutcliffe replied:
I'm fine with a breaking change, so long as we announce it clearly in the 1.3 release notes, and via the -dev mailing list when this is merged (so authors have time to update). Otherwise, perhaps we can catch self.set and redirect it?

dmitri-d replied:
Setting.self doesn't change DB state -- it only generates a hash that can be used to create a Setting object. The actual issue is in Setting#create code, and I can't simultaneously support both the old and the proposed call semantic. 

I think spreading the change over a couple of releases while not ideal is tolerable and leaves a copious amount of time for plugin authors to catch up. Having said that, I'm ok with announcing the breaking change and implementing it right away too.

domcleal replied:
I can't see why the breaking change is necessary to implement this - it looks like the same logic should fit into the existing methods.

That said, changing this is desirable at some point and I'd prefer to do it once rather than spread over multiple releases.  Is it possible to add a new interface and deprecate the old one for one release?

lzap replied:
> That said, changing this is desirable at some point and I'd prefer to do it once rather than spread over multiple releases.  Is it possible to add a new interface and deprecate the old one for one release?

Word.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

dmitri-d replied:
The issue is with the original implementation as it overrides create and create! methods. If I reset "readonly" flag in those, then the semantics of the call would allow to overwrite "default" attribute at any time.

I can keep Setting.set method around though -- this way way the old code would _mostly_ work, with the exception of updating of "default" attribute on startup.

As a side note, I think we should switch to migrations for updating of default values of settings -- the current approach results in a hacky workaround (among other things).

domcleal replied:
Sorry, I'm not sure I fully understand your argument @witlessbird.  It appears to me that it's by design that the default might change at any point in time.  Many of the settings used to get defaults from Puppet, so it'd "track" Puppet's settings, though it's not really the case any more.

dmitri-d replied:
@domcleal: my argument is that I cannot keep the old logic ("default" can be updated at any time) and enforce the new ("default" can only be updated on startup) at the same time.

domcleal replied:
Isn't it the value that's we're trying to update in this PR, rather than the default?

dmitri-d replied:
That wasn't my original understanding. If you think that the original semantic is no longer needed, wouldn't a real read-only Setting be better than a mostly read-only one (simpler implementation, no changes in original call semantic, etc).

domcleal replied:
Agreed I think, not updating the defaults makes more sense now that we've removed the Puppet dependency and so settings are pretty much static.  One related piece of information is that in #772 I was trying to update descriptions too, but that hasn't been merged for much the same reason as we're discussing now.

My main concern right now is keeping the current API functional, as I don't wish to stop plugins working this close to a release.

dmitri-d replied:
Somewhat agree re: concerns. I think there are two potential courses of action: a) ignore updates to default values when used with create/create! methods. I suspect these rarely change now, if at all. b) keep both, announce deprecation, and remove the old code in the next release.

domcleal replied:
If you did (b), what would you replace it with?  A new API?

dmitri-d replied:
a) switch to the current implementation: remove custom create and create! methods, use init_on_startup! to update default values. 
b) drop partial updates on startup altogether (config-file settings become fully read-only) and use migrations to update default values.

My preferred solution would be b).

domcleal commented on app/models/setting.rb:
```diff
@@ -200,6 +199,10 @@ def clear_cache
     Setting.cache.delete(name.to_s)
   end
 
+  def readonly_when_overriden_in_SETTINGS
```
Two 'd's in overridden :)

domcleal commented:
Closing in favour of #879, without the refactor.
